cpp/src/DataStorm/SessionI.cpp

```diff
@@ -743,8 +755,14 @@ SessionI::retry(NodePrx node, exception_ptr exception)
             return false;
         }
 
-        _retryTask = make_shared<IceInternal::InlineTimerTask>([node = std::move(node), self = shared_from_this()]
-                                                               { self->reconnect(node); });
+        _retryTask = make_shared<IceInternal::InlineTimerTask>(
+            [node = std::move(node), self = shared_from_this()]
+            {
+                self->_nextRetryAttemptTask =
+                    make_shared<IceInternal::InlineTimerTask>([self, node] { self->retry(node, nullptr); });
+                self->_instance->scheduleTimerTask(self->_nextRetryAttemptTask, 10s);
+                self->reconnect(node);
+            });
         _instance->scheduleTimerTask(_retryTask, delay);
```

Comment:
One gotcha here is we accumulate both delays, for example if this retry corresponds to nextRetryAttemptTask firing, we have already waited 10s, and now we will "delay" until reconnect is scheduled. I think will be good to discount the time we have already waited.